test/unit/coupling/interface/impl/SimpleMD/SimpleMDMoleculeTest.cpp

```diff
@@ -0,0 +1,67 @@
+#include <cppunit/TestFixture.h>
+#include <cppunit/extensions/HelperMacros.h>
+
+#include "coupling/interface/impl/SimpleMD/SimpleMDMolecule.h"
+
+#include "simplemd/Molecule.h"
+#include "tarch/la/Vector.h"
+
+#include <sstream>
+
+class SimpleMDMoleculeTest : public CppUnit::TestFixture
+{
+	CPPUNIT_TEST_SUITE(SimpleMDMoleculeTest);
+	CPPUNIT_TEST(testVelocity);
+	CPPUNIT_TEST(testForce);
+	CPPUNIT_TEST_SUITE_END();
+
+public:
+	void setUp()
+	{
+		
+	}
+	void tearDown()
+	{
+
+	}
+	void testVelocity()
+	{
+		simplemd::Molecule myMolecule;
+		coupling::interface::SimpleMDMolecule simpleMolecule(&myMolecule);
+		for(int i = -4000; i < 4001; i+=100)
+		{
+			for(int j = -4000; j < 4001; j+=100)
+			{
+				for(int k = -4000; k < 4001; k+=100)
+				{
+					tarch::la::Vector<3,double> velocity{i+0.5,j+0.5,k+0.5};
+
+					simpleMolecule.setVelocity(velocity);
+					tarch::la::Vector<3,double> storedVelocity = simpleMolecule.getVelocity();
+					CPPUNIT_ASSERT_MESSAGE( "velocity assersion",storedVelocity == velocity );
```

Comment:
I think it's okay to make that assumption since `la::Vector`'s overloaded `==` operator has the signature `template <int size, class T> bool operator==(const Vector<size, T>& l, const Vector<size, T>& r)` and internally it's doing the same thing anyway, so the assumption would be that the `==` operator in `la::Vector` is tested. Do you think that's valid? If not, I'll use `CPPUNIT_ASSERT_DOUBLES_EQUAL`, no problem

Reply:
Maybe my comment was not clear. I try again. Your code was assuming that

1.  `simpleMolecule.setVelocity()` accepts a `Vector<3,double>`. This is fine, since this is `SimpleMDMoleculeTest` and `simpleMolecule` is an `interface::SimpleMDMolecule`. So we definitely want to make assumptions about `simpleMolecule` and test them here.
2. `simpleMolecule.getVelocity()` returns a `Vector<3,double>`. Also fine.
3. `Vector<3,double>::operator==()` works as expected and compares the components using the `==` operator. Also fine, this should be tested for `la::Vector` elsewhere, here you can just assume that it works. 

So all of this is fine. My point was:

- By asserting that `storedVelocity == velocity`, since `simpleMolecule` is a wrapper around `simplemd::Molecule myMolecule` you implicitly also assumed that `myMolecule` internally has something like `Vector<3,double> _velocity` (or e.g. `double _velX, _velY, _velZ`). But `_velocity` is `private`.
- If we would change simplemd/Molecule.h to use e.g. `Vector<3,float> _velocity` or to use fixed-point math internally (keeping the signatures of all public methods constant), then the `SimpleMDMoleculeTest` would fail.
- I do not think that this test should fail in this case, since here you want to test `coupling::interface::SimpleMDMolecule` but not `simplemd::Molecule`, so this test should be independent of the implementation details of `simplemd::Molecule`.

Reply:
But the `simpleMolecule.getVelocity()` would always return a `Vector<3,double>` even if internally the `simplemd::Molecule` uses float or something else. I don't see the problem here.

Reply:
I see what Piet means
Let's say that the `simplemd::Molecule` internally converts `double` to `float`, and we lose some precision. Let's say that `getVelocity()` upcasts the `float` to `double` and returns it. If we check for equality at bit level, our tests will fail. But if we check numerical representation with a tolerance of 1e-6, it might still pass.

So the question is, what do we want this test to test? If we want it to enforce that absolutely nothing changes at bitlevel, then my way works. But if we want the `simplemd::Molecule` class to have the flexibility to do whatever it wants, as long as it gives us 6 digits after the decimal, then we do it Piet's way.

I'll change it to be Piet's way, then we can look at it further.